lib/internal/fs/copy/copy-sync.js

```diff
@@ -0,0 +1,334 @@
+'use strict';
+
+// This file is a modified version of the fs-extra's copySync method
+// modified in the following ways:
+//
+// - Use of the graceful-fs has been replaced with fs.
+// - Helpers from stat.js have been inlined.
+// - Formatting and style changes to match Node.js' linting rules.
+// - Parameter validation has been added.
+
+const { areIdentical, isSrcSubdir } = require('internal/fs/copy/copy');
+const { codes } = require('internal/errors');
+const {
+  os: {
+    errno: {
+      EEXIST,
+      EISDIR,
+      EINVAL,
+      ENOTDIR,
+    }
+  }
+} = internalBinding('constants');
+const {
+  ERR_FS_COPY_DIR_TO_NON_DIR,
+  ERR_FS_COPY_EEXIST,
+  ERR_FS_COPY_FIFO_PIPE,
+  ERR_FS_COPY_NON_DIR_TO_DIR,
+  ERR_FS_COPY_SOCKET,
+  ERR_FS_COPY_SYMLINK_TO_SUBDIRECTORY,
+  ERR_FS_COPY_TO_SUBDIRECTORY,
+  ERR_FS_COPY_UNKNOWN,
+} = codes;
+const fs = require('fs');
+const {
+  chmodSync,
+  closeSync,
+  copyFileSync,
+  existsSync,
+  lstatSync,
+  mkdirSync,
+  openSync,
+  readdirSync,
+  readlinkSync,
+  statSync,
+  symlinkSync,
+  unlinkSync,
+  utimesSync,
+} = fs;
+const path = require('path');
+const {
+  dirname,
+  join,
+  parse,
+  resolve,
+} = path;
+
+function copySyncFn(src, dest, opts) {
+  // Warn about using preserveTimestamps on 32-bit node
+  if (opts.preserveTimestamps && process.arch === 'ia32') {
+    const warning = 'Using the preserveTimestamps option in 32-bit' +
+      'node is not recommended';
+    process.emitWarning(warning, 'TimestampPrecisionWarning');
+  }
+  const { srcStat, destStat } = checkPathsSync(src, dest, opts);
+  checkParentPathsSync(src, srcStat, dest, 'copy');
+  return handleFilterAndCopy(destStat, src, dest, opts);
+}
+
+function checkPathsSync(src, dest, opts) {
+  const { srcStat, destStat } = getStatsSync(src, dest, opts);
+
+  if (destStat) {
+    if (areIdentical(srcStat, destStat)) {
+      throw new ERR_FS_COPY_TO_SUBDIRECTORY({
+        code: 'EINVAL',
+        message: 'src and dest cannot be the same directory',
```

Comment:
Change to `src and dest cannot be the same path`?

Reply:
That works but technically there might be cases where src and dest `path` are not the same but they are actually the same file.

- what's wrong with `src and dest cannot be the same`? Isn't descriptive enough?
- if we want to be explicit, we could check `srcStat.isDirectory()`, and if it's directory we say `src and dest cannot be the same directory` otherwise `src and dest cannot be the same file`.

Reply:
I'm happy with `src and dest cannot be the same`.

Reply:
I'm happy with that too.